# HG changeset patch
# User Rik <octave@nomad.inbox5.com>
# Date 1323134015 28800
#      Mon Dec 05 17:13:35 2011 -0800
# Node ID 15400d5deb1c249d44b54774b66011095c9710d8
# Parent  1221086f1ba56eb7f78e82ea27a3234aed1a0c84
ls.m: Fix handling of special characters in filename arguments (Bug #34950)

* ls.m: Escape special characters such as ["'(] in filename arguments.

diff --git a/scripts/miscellaneous/ls.m b/scripts/miscellaneous/ls.m
--- a/scripts/miscellaneous/ls.m
+++ b/scripts/miscellaneous/ls.m
@@ -46,24 +46,27 @@ function retval = ls (varargin)
     ls_command ();
   endif
 
   if (iscellstr (varargin))
 
     args = tilde_expand (varargin);
 
     if (nargin > 0)
-      ## FIXME -- this will fail for filenames that contain single quote
-      ## characters...
-      cmd = sprintf (" '%s'", args{:});
+      if (ispc () && ! isunix ())
+        ## shell (cmd.exe) on MinGW uses '^' as escape character
+        args = regexprep (args{:}, '([^\w.*?])', '^$1');
+      else
+        args = regexprep (args{:}, '([^\w.*?])', '\$1');
+      endif
     else
-      cmd = "";
+      args = "";
     endif
 
-    cmd = sprintf ("%s%s", __ls_command__, cmd);
+    cmd = sprintf ("%s %s", __ls_command__, args);
 
     if (page_screen_output () || nargout > 0)
 
       [status, output] = system (cmd);
 
       if (status == 0)
         if (nargout == 0)
           puts (output);
